client/state/reducer-registry/README.md

```diff
@@ -0,0 +1,26 @@
+Reducer Registry
+================
+
+The reducer registry allows us to load additional reducers into an original reducer, created by the usual `combineReducers`. It does that by implementing a special version of the same function, named `combineReducersAndAddLater`.
+
+To load additional reducers, call `addReducers` and pass the object containing the reducers, with the registry name already set.
+
+A little redux enhancer is implemented for the sole purpose of getting a hold of the global Redux store.
+
+To work with persistent states, `combineReducersAndAddLater` keeps the state tree inside a `_initialState` property when it couldn't find a reducer for the key, as opposed to removing the state altogether. When the reducer matching the key is loaded, the serialized state held by the property will get passed into the loaded reducer so that it can be properly deserialized, and/or validated with the schema. This is achieved by calling a placeholder reducer whenever a reducer matching the key is unfound. The same placeholder reducer also properly "serialize" the state by returning the state held by the `_initialState` property.
```

Comment:
What if we created additional keys in browser storage?

Reply:
@coderkevin what have persistence needs looked like so far for extensions? I'm not sure how much of our redux data we're going to eventually expose, so we might even have a system like a key/value store that might be more useful.

Reply:
@gwwar We might not want to split reducers at the top level, and we might not want to couple the way storage splits up w/ how we split up reducers. The pull request here only deals with extensions reducers, but I would like to use the same mechanics to split up core reducers as we see fit.

What's the reason behind splitting up browser storage? Have we already found performance issues etc with the current setup?

Reply:
Yes, hydrating state is pretty slow, especially as our global state grows. We also have added validation checks to throw out old data. (We could in the future change this to a faster version check but it'd take more work). See a previous exploration in https://github.com/Automattic/wp-calypso/pull/9957

Looking back at the parent issue, I think we're more interested with if we can add and remove reducers async which is a very cool idea. For now, I'd suggest we take a step back and ignore persistence concerns for this first step, as its an optional enhancement that should always be carefully evaluated before its put in use.

Reply:
Thanks for the pointer. I will leave this part of the code as-is.

Reply:
Is part of the slow state hydration due to the time it takes to load data from IndexedDB? I'd love to know so I could see about anything localForage can do to improve those first-load calls.

In general first-load performance on IndexedDB in Firefox and Chrome has not been amazing–it certainly used to be the case that because localStorage data for a page was loaded and cached in-memory on page load using localStorage was *faster* for one-time loading of a simple data-set (not sure if https://www.peterbe.com/plog/localstorage-is-fast is still the case but it certainly was a few years ago).

That PR mentions issues with Safari accessing IndexedDB from a web worker, but that has since been resolved.

Side note: I'd like to focus on common performance issues in 2.x versions of localForage; while some of that is likely up to the underlying IndexedDB implementations, I'll have a look at calypso's usage and see if there's anything we're doing on the localForage side that's slowing things down we could improve.

`</end-thread-hijack-by-localForage-brain>`

Reply:
@tofumatt I don't think localForage is the bottleneck. It's been a while since I've done a profiling pass but mainly we have a lot of state that may be persisted (currently unbounded), and we also use JSON schemas to throw out older versions of data (slow) vs a numbered versioning system (hard to enforce/keep up to date). Since we store global state in a single key, it also becomes a little difficult to balance how often we persist to how less useful it is if we do it less often. Some improvements for Calypso persistence would be better cache eviction / splitting persisted state into more keys so we don't need to write all persisted state at once.

Reply:
> what have persistence needs looked like so far for extensions?

So far for Store, we've not used it for now until a later date. I see this kind of persistence as a performance enhancement and not one to be used lightly as the data storage implications can be large and far-reaching.

Reply:
> not one to be used lightly as the data storage implications can be large and far-reaching

💯